p2p/src/network/kad/request/p2p_network_kad_request_reducer.rs

```diff
@@ -1,52 +1,293 @@
+                    })
+                    .and_then(|(P2pNetworkConnectionMuxState::Yamux(yamux), incoming)| {
+                        yamux
+                            .next_stream_id(crate::YamuxStreamKind::Kademlia, incoming)
+                            .ok_or_else(|| format!("cannot get next stream for {addr}"))
+                    })?;
+
+                dispatcher.push(P2pNetworkYamuxAction::OpenStream {
+                    addr: *addr,
+                    stream_id,
+                    stream_kind: crate::token::StreamKind::Discovery(
+                        crate::token::DiscoveryAlgorithm::Kademlia1_0_0,
+                    ),
+                });
+                dispatcher.push(P2pNetworkKadRequestAction::StreamIsCreating {
+                    peer_id: *peer_id,
+                    stream_id,
+                });
+                Ok(())
             }
-            P2pNetworkKadRequestAction::MuxReady { .. } => {}
             P2pNetworkKadRequestAction::StreamIsCreating { stream_id, .. } => {
-                self.status = super::P2pNetworkKadRequestStatus::WaitingForKadStream(*stream_id)
+                request_state.status = P2pNetworkKadRequestStatus::WaitingForKadStream(*stream_id);
+
+                Ok(())
             }
-            P2pNetworkKadRequestAction::StreamReady { .. } => {
-                let find_node =
-                    P2pNetworkKademliaRpcRequest::find_node(self.key).map_err(|e| e.to_string())?;
+            P2pNetworkKadRequestAction::StreamReady {
+                peer_id,
+                stream_id,
+                addr,
+            } => {
+                let find_node = P2pNetworkKademliaRpcRequest::find_node(request_state.key)
+                    .map_err(|e| e.to_string())?;
+
                 let message = super::super::Message::from(&find_node);
-                self.status = quick_protobuf::serialize_into_vec(&message).map_or_else(
+                request_state.status = quick_protobuf::serialize_into_vec(&message).map_or_else(
                     |e| {
                         super::P2pNetworkKadRequestStatus::Error(format!(
                             "error serializing message: {e}"
                         ))
                     },
                     super::P2pNetworkKadRequestStatus::Request,
                 );
+                let peer_id = *peer_id;
+                let stream_id = *stream_id;
+                let addr = *addr;
+                let key = request_state.key;
+
+                let dispatcher = state_context.into_dispatcher();
+                let data =
+                    P2pNetworkKademliaRpcRequest::find_node(key).map_err(|e| e.to_string())?;
+                dispatcher.push(P2pNetworkKademliaStreamAction::SendRequest {
+                    addr,
+                    peer_id,
+                    stream_id,
+                    data,
+                });
+                dispatcher.push(P2pNetworkKadRequestAction::RequestSent { peer_id });
```

Comment:
@0xMimir does this action depend on the previous one? if that is the case, then it should be handled with a callback.

Doesn't need to be fixed in this PR but should at least be a TODO for it.

Reply:
relevant PR where I already updated some code using callbacks to enforce the dependency https://github.com/openmina/openmina/pull/628/files

Reply:
main reason is that if for some reason `P2pNetworkKademliaStreamAction::SendRequest` doesn't get dispatched (either gets cancelled or filtered out by the enabling conditions), we don't want to dispatch `P2pNetworkKadRequestAction::RequestSent` either, since it is a dependent action.

Reply:
@0xMimir also https://github.com/openmina/openmina/blob/develop/ARCHITECTURE.md#callbacks-instead-of-conditional-dispatches